### PR TITLE
Add API Change password

### PR DESCRIPTION
**Pull request by ntk148v**
Close: #105

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -119,6 +119,7 @@ func (a *API) Register(r *mux.Router) {
 	resr.HandleFunc("/users", a.addUser).Methods("OPTIONS", "POST")
 	resr.HandleFunc("/users/{user:[a-z 0-9]+}", a.removeUser).Methods("OPTIONS", "DELETE")
 	resr.HandleFunc("/users", a.listUsers).Methods("OPTIONS", "GET")
+	resr.HandleFunc("/users/{user:[a-z 0-9]+}/change_password", a.changePassword).Methods("OPTIONS", "PUT")
```
Comment (vtdat) [inline api/api.go:122]:
```suggestion
	resr.HandleFunc("/users/{user:[a-z 0-9]+}", a.changePassword).Methods("OPTIONS", "PUT")
```

this is more reasonable

File: api/users.go [hunk 2]
```diff
@@ -227,3 +227,74 @@ func (a *API) listUsers(w http.ResponseWriter, req *http.Request) {
...
 }
+
+// changePassword updates the new password for a given user
+func (a *API) changePassword(w http.ResponseWriter, req *http.Request) {
+	vars := mux.Vars(req)
+	username := vars["user"]
+	if username == config.Get().AdminAuthentication.Username {
+		a.respondError(w, apiError{
+			code: http.StatusForbidden,
+			err:  errors.New("Can not change the administrator user's password"),
```
Comment (vtdat) [inline api/users.go:238]:
```suggestion
			err:  errors.New("Cannot change administrator's password"),
```

File: api/users.go [hunk 3]
```diff
@@ -227,3 +227,74 @@ func (a *API) listUsers(w http.ResponseWriter, req *http.Request) {
...
+		return
+	}
+	if err := req.ParseForm(); err != nil {
+		a.respondError(w, apiError{
+			code: http.StatusBadRequest,
+			err:  err,
+		})
+		return
+	}
+	newpassword := req.Form.Get("newpassword")
```
Comment (vtdat) [inline api/users.go:249]:
```suggestion
	password := req.Form.Get("password")
```

File: api/users.go [hunk 4]
```diff
@@ -227,3 +227,74 @@ func (a *API) listUsers(w http.ResponseWriter, req *http.Request) {
...
+	if newpassword == "" {
```
Comment (vtdat) [inline api/users.go:250]:
```suggestion
	if password == "" {
```

File: api/users.go [hunk 5]
```diff
@@ -227,3 +227,74 @@ func (a *API) listUsers(w http.ResponseWriter, req *http.Request) {
...
+	}
+	if len(resp.Kvs) == 0 {
+		a.respondError(w, apiError{
+			code: http.StatusBadRequest,
+			err:  errors.New("Unknown user"),
+		})
+		return
+	}
+	// Hash for new password
+	hashed, err := common.GenerateBcryptHash(newpassword, config.Get().PasswordHashingCost)
```
Comment (vtdat) [inline api/users.go:275]:
```suggestion
	hashedpw, err := common.GenerateBcryptHash(password, config.Get().PasswordHashingCost)
```

File: api/users.go [hunk 6]
```diff
@@ -227,3 +227,74 @@ func (a *API) listUsers(w http.ResponseWriter, req *http.Request) {
...
+		a.respondError(w, apiError{
+			code: http.StatusInternalServerError,
+			err:  errors.Wrap(err, "Something went wrong"),
+		})
+		return
+	}
+
+	user := &model.User{
+		Username: username,
+		Password: hashed,
```
Comment (vtdat) [inline api/users.go:286]:
```suggestion
		Password: hashedpw,
```

File: docs/authn-authz.md [hunk 7]
```diff
@@ -89,6 +90,16 @@ Authorization: Bearer <token>
 | --------- | ----- | ------ | -------- | ------- | ----------- |
 | name      | query | string | true     |         | User's name |
 
+### 2.5. Change user password
+
+**PATH**: `/users/{user}/change_password`
```
Comment (vtdat) [inline docs/authn-authz.md:95]:
```suggestion
**PATH**: `/users/{user}`
```

File: docs/authn-authz.md [hunk 8]
```diff
@@ -89,6 +90,16 @@ Authorization: Bearer <token>
...
+**METHOD**: `PUT`
+
+| Parameter   | In    | Type   | Required | Default | Description         |
+| ----------- | ----- | ------ | -------- | ------- | ------------------- |
+| user        | path  | string | true     |         | User's name         |
+| newpassword | query | string | true     |         | User's new password |
```
Comment (vtdat) [inline docs/authn-authz.md:101]:
```suggestion
| user        | path  | string | true     |         | Username            |
| password    | query | string | true     |         | User's new password |
```

File: api/users.go (hunk 4 above)
Comment (vtdat) [inline api/users.go:250]:
```suggestion
	if password == "" {
```

File: api/api.go (hunk 1 above)
Comment (ntk148v) [reply api/api.go:122]: > this is more reasonable

IMO, we should specify the **action**, in this case - `change password`.

**Pull request merged**